### ISSUE TITLE
Privilege Separation

### ISSUE THREAD

**Issue by ollieparanoid** [postmarketOS/pmbootstrap#846]
@ata2001 asked in the chat:
> any idea, how are we going to implement privilege separation? like every app should run as its own user/group and have its own home folder

I have put some thought into this topic already and would like to share my opinion. For me, privilege separation is essential to have a reasonable secure system. All software has bugs, so we should make sure that these bugs have little to zero impact to the security of the whole system, and this is made possible with privilege separation.

My approach would be using [bubblewrap](https://github.com/projectatomic/bubblewrap) and/or [firejail](https://github.com/netblue30/firejail) wrappers for at least the most popular programs. Both use seccomp filters from the kernel. We could put these wrappers into `/usr/lib/hardening-wrapper/bin` or a similar path (that's where Arch Linux ships wrappers for GCC to enable hardening flags by default) and they would use bubblewrap/firejail to start the actual program, but with disabled network or file access, or other syscalls - depending on what makes sense for the program.

We could put all wrappers into a `postmarketos-hardening` package or something, and use `install_if` rules to install only the wrappers for the programs that are actually installed.

With that being said, there's also SELinux and I can't really say what the advantages/disadvantages would be, so we will need more research and more opinions on that.

**Comment by lawl** [comment]
If I can quickly leave my opinion here which should mean absolutely nothing btw. I looked at the code for firejail and to put it delicately, I wasn't impressed. Besides the code being of questionable quality, they use OverlayFS which doesn't exactly have a [a good track record](https://www.rapid7.com/db/modules/exploit/linux/local/overlayfs_priv_esc) of being useful for security stuff. 

I heard of bubblewrap for the first time here. But I definitely agree that seccomp+namespaces are the way to go. Just, maybe reconsider using [firejail, which implements X11 sandboxing and than fails at exactly that](http://archive.is/QbGV9).

With that said I'd just like to throw two more links to look at. Google's [nsjail](https://github.com/google/nsjail) and [oz](https://github.com/subgraph/oz).

Note: I'm not exactly an expert on this topic, but some time ago, I started working on my [own solution to this problem](https://github.com/lawl/qbit). But ended up deciding it wasn't worth it because oz and nsjail both looked reasonable to me. Just... maybe reconsider using vanilla firejail.

**Comment by MartijnBraam** [comment]
I don't think we need containerisation at all for privilege seperation, thats what selinux is for. we just need to ship with selinux config for the applications.

I haven't researched this very deeply but it looks like with seccomp you can disable specific syscals but with selinux you can give the software a security context and then allow/disallow paths and networking stuff.

Depending on containers (namespaces) in linux for security isn't a good idea since not everything in linux has namespace support and is just reachable from inside the container.
